File: tree/java/PathSum.java
package com.leetcode.tree;

import java.util.ArrayList;
import java.util.List;

public class GetPathSum {
    public static void main(String[] args) {
        TreeNode root = new TreeNode(5);
        TreeNode node1 = new TreeNode(4);
        TreeNode node2 = new TreeNode(8);

        TreeNode node3 = new TreeNode(11);

        TreeNode node4 = new TreeNode(13);
        TreeNode node5 = new TreeNode(4);
        TreeNode node6 = new TreeNode(7);
        TreeNode node7 = new TreeNode(2);
        TreeNode node8 = new TreeNode(5);
        TreeNode node9 = new TreeNode(1);

        root.left = node1;
        root.right = node2;
        node1.left = node3;
        node2.left = node4;
        node2.right = node5;
        node3.left = node6;
        node3.right = node7;
        node5.left = node8;
        node5.right = node9;

        System.out.println(pathSum(root, 22));
    }


    public static List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> all = new ArrayList<>();
        hasPathSum(root, sum, all, new ArrayList<>());
        return all;
    }


    public static void hasPathSum(TreeNode root, int sum, List<List<Integer>> all, List<Integer> curList) {
        if (root == null) {
            return;
        }

        if (root.left == null && root.right == null) {
            if (root.val == sum) {
                ArrayList<Integer> arrayList = new ArrayList<>(curList);
                arrayList.add(root.val);
                all.add(new ArrayList<>(arrayList));
            }
            return;
        }
        curList.add(root.val);
        hasPathSum(root.left, sum - root.val, all, curList);
        hasPathSum(root.right, sum - root.val, all, curList);
        if (curList.size() > 0) curList.remove(curList.size() - 1);
    }
}

File: go/tree/path_sum.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func pathSum1(root *TreeNode, sum int) [][]int {
	var path = make([]int, 0)
	var paths = make([][]int, 0)
	hasPathSum1(root, sum, &paths, path)

	fmt.Print(paths)
	return paths
}

func hasPathSum1(root *TreeNode, sum int, paths *[][]int, path []int) {
	if root == nil {
		return
	}

	if root.Left == nil && root.Right == nil {
		if root.Val == sum {
			path = append(path, root.Val)
			*paths = append(*paths, path)
		}
		return
	}

	path = append(path, root.Val)

	hasPathSum1(root.Left, sum-root.Val, paths, path)
	hasPathSum1(root.Right, sum-root.Val, paths, path)

	//l := len(path)
	//path = append(path[:l-2])
}

func main() {
	root := &TreeNode{Val: -6}
	node1 := &TreeNode{Val: -3}
	node2 := &TreeNode{Val: -6}
	node3 := &TreeNode{Val: 0}

	node4 := &TreeNode{Val: -6}
	node5 := &TreeNode{Val: -5}
	node6 := &TreeNode{Val: 4}
	node7 := &TreeNode{Val: 1}
	node8 := &TreeNode{Val: 7}

	root.Right = node1
	node1.Left = node2
	node1.Right = node3

	node2.Left = node4
	node2.Right = node5

	node3.Left = node6

	node6.Left = node7
	node6.Right = node8

	pathSum1(root, -21)
}
